include/albatross/src/indexing/group_by.hpp

```diff
@@ -466,6 +466,36 @@ template <typename Derived> class GroupByBase {
     return output;
   }
 
+  template <typename SubsetableType,
+            typename std::enable_if<
+                details::is_subsetable<SubsetableType>::value, int>::type = 0>
+  Grouped<KeyType, std::pair<ValueType, SubsetableType>>
+  with(const SubsetableType &other) const {
+    assert(parent_.size() == other.size());
+    Grouped<KeyType, std::pair<ValueType, SubsetableType>> output;
+    for (const auto &key_indexer_pair : indexers()) {
+      output.emplace(
+          key_indexer_pair.first,
+          std::make_pair(albatross::subset(parent_, key_indexer_pair.second),
+                         albatross::subset(other, key_indexer_pair.second)));
+    }
+    return output;
+  }
+
+  template <template <typename...> class Map, typename AlreadyGroupedType>
+  Grouped<KeyType, std::pair<ValueType, AlreadyGroupedType>>
+  with(const Map<KeyType, AlreadyGroupedType> &other) const {
+    assert(other.size() == indexers().size());
+    Grouped<KeyType, std::pair<ValueType, AlreadyGroupedType>> output;
+    for (const auto &key_indexer_pair : indexers()) {
+      output.emplace(
```

Comment:
Do we want to assert that the size of each group between `this` and `other`? We'll fail if any group names don't align, but sizes within groups aren't checked. What happens if:
```
fib_group = group_by( fib(10) , number_of_digits)
prime_group = group_by( primes(10) , number_of_digits)
ill_defined = fib_group.with(prime_group)
```

Reply:
I think we specifically want to allow different sized groups.  For example, it could be that you first want to compute a map `std::map<Gender, double>` which contains the life expectancy by gender.  Then you might have a dataset and you want to compute the percentage of people over that life expectancy.  You'd want to do something like:
```
dataset.group_by(get_gender).with(life_expectancy).apply(percentage_over_expectancy);
```
If you look at the types you have `dataset.group_by(get_gender)` which would be a map from gender to a dataset. and you'd be joining it with a map from gender to a double.  In otherwords the with argument may not even have a `size()`.